p2p/net/swarm/dial_ranker.go

```diff
@@ -88,48 +99,100 @@ func DefaultDialRanker(addrs []ma.Multiaddr) []network.AddrDelay {
 // addresses relative to direct addresses.
 func getAddrDelay(addrs []ma.Multiaddr, tcpDelay time.Duration, quicDelay time.Duration,
 	offset time.Duration) []network.AddrDelay {
+	if len(addrs) == 0 {
+		return nil
+	}
 
 	sort.Slice(addrs, func(i, j int) bool { return score(addrs[i]) < score(addrs[j]) })
 
-	// If the first address is (QUIC, IPv6), make the second address (QUIC, IPv4).
-	happyEyeballs := false
-	if len(addrs) > 0 {
+	// addrs is now sorted by (Transport, IPVersion). Reorder addrs for happy eyeballs dialing.
```

Comment:
Should this sorting logic be part of `score`?

Reply:
I'm not sure if that'd be more readable. For this ranking we need to look at other addresses, so an IPv4 address gets higher score than other IPv6 addresses only if there are some IPv6 addresses present. I think the present implementation makes the logic explicit.